src/events/zookeeper.py

```diff
@@ -71,7 +71,7 @@ def _on_zookeeper_changed(self, event: RelationChangedEvent) -> None:
             event.defer()
             return
```

Comment:
**question**: I think we should pre-emptively stop the workload anytime we early return in this method. What do you think?

Reply:
Why do you suggest this? I am not sure if it's a good or bad idea yet, so wondering if there's some edge-case you found?

Reply:
I think I would rather keep it as is, now that I think about it.
If Kafka cannot connect to ZooKeeper, it'll error out and exit. The `systemd` service that the Snap creates will keep attempting to restart the service, until ZooKeeper can actually be reached.

You're probably right about all the deferred events though. I think when we have a bit more headspace, this would be a good candidate to remove some extraneous defers and stabalise the charm a bit more.

Reply:
I guess the 'setting of the flag' would be https://github.com/canonical/zookeeper-operator/pull/162/files#r1786750206

Reply:
Why don't you want to stop the service Marc? If there is no connection to Zk, there is no point on running Kafka. I'd rather avoid having logs flooded with errors that we expect.

Reply:
It's less that I don't want to stop it, it's that I don't want it to 'stop trying to restart through systemd'

Reply:
We also have to be careful not to couple the two charms too much.
If we imagine we were developing Kafka in a vacuum, and didn't have any awareness of what this 'restore' stuff was, we wouldn't just drop Kafka services immediately on `relation-changed`.

Reply:
That's a very fair point, especially considering KRaft coming up. Closing this thread, as I will not act upon it right now